server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -322,7 +325,13 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
   {
     try {
       log.info("Loading segment %s", segment.getIdentifier());
-      if (loadSegment(segment, callback)) {
+      if (segmentsToDelete.contains(segment)) {
+        synchronized (lock) {
+          segmentsToDelete.remove(segment);
+        }
+      }
+      loadSegment(segment, callback);
```

Comment:
This originates from the scenario where 
Step 1. ```addSegment()``` is called, and segment is announced.
Step 2. ```removeSegment()``` is called, segment is unannounced, a delete runnable is scheduled.
Step 3. ```addSegment()``` is called before the scheduled delete runnable executes, but it won't announce the segment because ```loadSegment()``` returns false(the segment is still loaded at this moment), since segment is removed from ```segmentsToDelete```, the scheduled delete runnable will become a noop, which makes this historical never be able to announce that segment.

The fix in this PR uses ```isAnnounced(segment)``` to check if a historical should announce/unannounce a segment.